src/entities/nestedJoin/index.ts

```diff
@@ -0,0 +1,270 @@
+import {
+    createPublicClient,
+    decodeAbiParameters,
+    decodeFunctionResult,
+    encodeFunctionData,
+    http,
+} from 'viem';
+import { Address, Hex } from '../../types';
+import { Slippage } from '../slippage';
+import { Token } from '../token';
+import { BALANCER_RELAYER, CHAINS, getPoolAddress } from '../../utils';
+import { Relayer } from '../relayer';
+import { parseNestedJoinArgs } from '../utils/parseNestedJoinArgs';
+import { TokenAmount } from '../tokenAmount';
+import { balancerRelayerAbi, bathcRelayerLibraryAbi } from '../../abi';
+
+export type NestedJoinInput = {
+    amountsIn: {
+        address: Address;
+        rawAmount: bigint;
+    }[];
+    chainId: number;
+    rpcUrl: string;
+    testAddress: Address;
+    useNativeAssetAsWrappedAmountIn?: boolean;
+    fromInternalBalance?: boolean;
+};
+
+export type JoinStep = {
+    action: 'join'; // TODO: are we supporting other actions, such as swaps?
+    level: number; // 0 is the bottom and the highest level is the top
+    input: NestedToken;
+    poolId: Address; // output poolId
+    isTop: boolean;
+};
+
+export type NestedPoolState = {
+    pools: {
+        id: Hex;
+        address: Address;
+        type: string;
+    }[];
+    joinSteps: JoinStep[]; // each token should have at least one
+};
+
+type NestedToken = {
+    address: Address;
+    decimals: number;
+    index: number;
+};
+
+export type NestedJoinArgs = {
+    chainId: number;
+    useNativeAssetAsWrappedAmountIn?: boolean;
+    sortedTokens: Token[];
+    poolId: Address;
+    poolType: string;
+    kind: number;
+    sender: Address;
+    recipient: Address;
+    maxAmountsIn: {
+        amount: bigint;
+        isRef: boolean;
+    }[];
+    minBptOut: bigint;
+    fromInternalBalance: boolean;
+    outputReferenceKey: bigint;
+};
+
+type NestedJoinQueryResult = {
+    calls: NestedJoinArgs[];
+    bptOut: TokenAmount;
+};
+
+type NestedJoinCallInput = NestedJoinQueryResult & {
+    slippage: Slippage;
+    sender: Address;
+    recipient: Address;
+    relayerApprovalSignature?: Hex;
+};
+
+const outputRefOffset = 100n;
+
+export class NestedJoin {
+    async query(
```

Comment:
I think in general this function could be refactored to smaller functions that will make it clearer, e.g.  `getPoolCall`, `getMaxAmountsIn()`, etc

Reply:
Totally agreed. I thought about leaving this kind of refactor to after adding more test scenarios, because I think the code will change a bit until then.